### DISCUSSION TITLE
Replacing whitespace with underscores on the fly with an insert node

### DISCUSSION THREAD

**Discussion by dbatten5** [L3MON4D3/LuaSnip#1096] [Q&A]
One of my favourite snippets I used with ultisnips was this
```
snippet f
def $1`!p
t[1] = t[1].replace(' ', '_')
`($2) -> $3:
	$0
endsnippet
```
It meant that as I typed out the name of a function with spaces, the snippet replaced the spaces with underscores on the fly, meaning I can just freely type out the name of a snake_case function without having to insert underscores. I'm struggling to recreate this with LuaSnip. I figured it might be some combination of function nodes and insert nodes but for the life of me I can't get it to work. Any suggestions?

**Comment by L3MON4D3** [comment]
Ah, yeah, that's still a WIP :sweat_smile:
We have dynamicNode, but it can't yet depend on a node inside itself.. but your question sparked some ideas for me, and I'm pretty optimistic that those will bear fruit :)

**Reply by dbatten5** [reply]
Great to hear! I thought supplying some sort of callback that gets run on the output of a node would do the trick but sure you've got some better ideas

**Reply by L3MON4D3** [reply]
Oh yeah, you can also do that, but it's not as general as I'd like it to be :D
But it works!:

```lua
local events = {[1] = {
	[events.enter] = function(node)
		node.au_id = vim.api.nvim_create_autocmd("TextChangedI", {
			callback = function()
				local t = node:get_text()[1]
				local mod_t = t:gsub(" ", "_")
				node:set_text({mod_t})
			end
		})
	end,
	[events.leave] = function(node)
		vim.api.nvim_del_autocmd(node.au_id)
	end}
}
ls.add_snippets("all", {
	s("def", {
		t"def ", i(1, "function-name")
	}, {
		callbacks = events
	})
}, {key = "def"})
```

**Reply by dbatten5** [reply]
Ok nice, looking forward to seeing what you come up with 👍

**Reply by L3MON4D3** [reply]
Hey, check #1137, it should handle this case just fine :)

**Reply by dbatten5** [reply]
Hi @L3MON4D3 , thanks for messaging. Do you have any pointers on how to achieve it with self-dependent DynamicNodes? I've been trying to make sense of the example you linked but I'm not getting very far with it

**Comment by fovj** [comment]
![ezgif-7-5423f8bfc2](https://github.com/L3MON4D3/LuaSnip/assets/4456413/49619afb-dc2e-420f-a0c3-8c0e8bd5d5e4)

Like this?

**Reply by dbatten5** [reply]
Not quite, I'm looking to replace the output of the `InputNode` in place, not display the transformed output in another node

**Reply by fovj** [reply]
Oh, yes, I've encountered the same issue. https://github.com/L3MON4D3/LuaSnip/discussions/1102
It's possible to pass parameters like "transform" to `insertNode`.
